File: Parciales y simulacros/Python/Parcial 1C2024 Comision A/parcial.py
from queue import Queue as Cola

"""
Ejercicio 1: Atención por Guardia (1 punto)
Desde el Hospital Fernandez nos pidieron solucionar una serie de problemas relacionados con la información que
maneja sobre los pacientes y el personal de salud. En primer lugar debemos resolver en qué orden se deben atender
los pacientes que llegan a la guardia. En enfermería, hay una primera instancia que clasifica en dos colas a los pacientes:
una urgente y otra postergable (esto se llama hacer triage). 
A partir de dichas colas que contienen la identificación del paciente, se pide devolver una nueva cola según la siguiente especificación.

problema orden_de_atencion(in urgentes: cola⟨Z⟩, in postergables: cola⟨Z⟩) : cola⟨Z⟩ {
  requiere: {no hay elementos repetidos en urgentes}
  requiere: {no hay elementos repetidos en postergables}
  requiere: {la intersección entre postergables y urgentes es vacía}
  requiere: {|postergables| = |urgentes|}
  asegura: {no hay repetidos en res }
  asegura: {res es permutación de la concatenación de urgentes y postergables}
  asegura: {Si urgentes no es vacía, en tope de res hay un elemento de urgentes}
  asegura: {En res no hay dos seguidos de urgentes}
  asegura: {En res no hay dos seguidos de postergables}
  asegura: {Para todo c1 y c2 de tipo "urgente" pertenecientes a urgentes si c1 aparece antes 
            que c2 en urgentes entonces c1 aparece antes que c2 en res}
  asegura: {Para todo c1 y c2 de tipo "postergable" pertenecientes a postergables si c1 aparece 
            antes que c2 en postergables entonces c1 aparece antes que c2 en res}
"""
def orden_de_atencion(urgentes: Cola[int], postergables: Cola[int]) -> Cola[int]:
    return


"""
Ejercicio 2: Alarma epidemiológica. (3 puntos)
Necesitamos detectar la aparición de posibles epidemias. Para esto contamos con un lista de enfermedades infecciosas 
y los registros de atención por guardia dados por una lista expedientes. 
Cada expediente es una tupla con ID paciente y enfermedad que motivó la atención. 
Debemos devolver un diccionario cuya clave son las enfermedades infecciosas y su valor es la proporción de pacientes 
que se atendieron por esa enfermedad.
En este diccionario deben aparecer solo aquellas enfermedades infecciosas cuya proporción supere cierto umbral.

problema alarma_epidemiologica (registros: seq⟨ZxString⟩, infecciosas: seq⟨String⟩, umbral: R) : dict⟨String,R⟩ {
  requiere: {0 < umbral < 1}
  asegura: {claves de res pertenecen a infecciosas}
  asegura: {Para cada enfermedad perteneciente a infecciosas, si el porcentaje de pacientes que se atendieron por esa enfermedad 
            sobre el total de registros es mayor o igual al umbral, entonces res[enfermedad] = porcentaje}
  asegura: {Para cada enfermedad perteneciente a infecciosas, si el porcentaje de pacientes que se atendieron por esa enfermedad 
            sobre el total de registros es menor que el umbral, entonces enfermedad no aparece en res}
}
"""

def alarma_epidemiologica(registros: list[tuple[int, str]], infecciosas: list[str], umbral: float) -> dict[str, float]:
    res: dict[str, float] = {}
    return res 


"""
Ejercicio 3: Empleados del mes (2 puntos)
Dado un diccionario con la cantidad de horas trabajadas por empleado, en donde la clave es el ID del empleado
y el valor es una lista de las horas trabajadas por día, queremos saber quienes trabajaron más para darles un premio.
Se deberá buscar la o las claves para la cual se tiene el máximo valor de cantidad total de horas, y devolverlas en una lista.

problema empleados_del_mes(horas:dicc⟨Z,seq⟨Z⟩⟩) : seq⟨Z⟩ {
  requiere: {No hay valores en horas que sean listas vacías}
  asegura: {Si ID pertence a res entonces ID pertence a las claves de horas}
  asegura: {Si ID pertenece a res, la suma de sus valores de horas es el máximo de la suma de elementos de horas de todos los otros IDs}
  asegura: {Para todo ID de claves de horas, si la suma de sus valores es el máximo de la suma de elementos de horas de los otros IDs, 
            entonces ID pertences a res}
}
"""

def empleados_del_mes(horas: dict[int, list[int]]) -> list[int]:
    res: list[int] = []
    return res


"""
Ejercicio 4: Nivel de ocupación del hospital (2 puntos)
Queremos saber qué porcentaje de ocupación de camas hay en el hospital. El hospital se representa por una matriz en donde las filas 
son los pisos, y las columnas son las camas. Los valores de la matriz son booleanos que indican si la cama está ocupada o no. 
Si el valor es verdadero (True) indica que la cama está ocupada. 
Se nos pide programar en Python una función que devuelve una secuencia de reales, indicando la proporción de camas ocupadas en cada piso.

problema nivel_de_ocupacion(camas_por_piso:seq⟨seq⟨Bool⟩⟩) : seq⟨R⟩ {
  requiere: {Todos los pisos tienen la misma cantidad de camas.}
  requiere: {Hay por lo menos 1 piso en el hospital.}
  requiere: {Hay por lo menos una cama por piso.}
  asegura: {|res| = |camas_por_piso|}
  asegura: {Para todo 0<= i < |res| se cumple que res[i] es igual a la cantidad de camas ocupadas del piso i dividido el total de camas del piso i)}

"""
def nivel_de_ocupacion(pisos: list[list[bool]]) -> list[float]:
    res: list[float] = []
    return res

"""
Ejercicio 5: Preguntas teóricas (2 puntos)
Conteste marcando la opción correcta.

A) ¿Qué es una variable con 'scope local' en Python? (0.75 puntos)

 1) Una variable definida fuera de cualquier función y accesible en todo el programa.
 2) Una variable definida dentro de una función, que solo puede ser utilizada dentro de esa función.
 3) Una variable que puede ser utilizada en cualquier módulo importado.


B) ¿Qué es una estructura de control en Python? (0.75 puntos)

 1) Una herramienta que permite la ejecución condicional y repetitiva de bloques de código.
 2) Un tipo especial de variable que almacena datos complejos.
 3) Una librería utilizada para manipular archivos en el sistema operativo.


C) ¿Qué representa un nodo en un Control Flow Graph? (0.5 puntos)

 1) Una variable utilizada en el programa.
 2) Una condición lógica o una instrucción en el código.
 3) Un archivo de datos externo
"""

File: Parciales y simulacros/Haskell/Parcial 1C2024 Comision A/parcial.hs
module Parcial where
{-Enunciado
La codificación por sustitución es una de las técnicas de cifrado más simples, en el que un caracter en el texto original es 
reemplazado por otro caracter dependiendo de un mapeo. Este mapeo puede representarse con una secuencia de tuplas de dos caracteres,
donde la primera componente de la tupla representa el caracter original y la segunda componente el caracter por el cual se lo va a 
sustituir. Por simplicidad, en este problema codificaremos solo los caracteres que aparecen en el mapeo dado. 
Todos los restantes caracteres quedan inalterados en el mensaje codificado.
Para implementar este sistema de codificación nos enviaron las siguientes especificaciones y nos pidieron que hagamos el desarrollo
enteramente en Haskell, utilizando los tipos requeridos y solamente las funciones que se ven en la materia 
Introducción a la Programación / Algoritmos y Estructuras de Datos I (FCEyN-UBA).
-}

{-Ejercicio 1 (2 puntos)
problema hayQueCodificar (c: Char, mapeo: seq⟨Char x Char⟩ ) : Bool {
  requiere: {No hay elementos repetidos entre las primeras componentes de mapeo}
  requiere: {No hay elementos repetidos entre las segundas componentes de mapeo}
  asegura: {res = true <=> c es igual a la primera componente de alguna tupla de mapeo}
}
-}
longitud :: (Eq t) => [t] -> Int 
longitud lista | lista == [] = 0
               | otherwise = 1 + longitud(tail lista)

-- pertenece :: (Eq t) => [t] -> t -> Bool
-- pertenece lista elemento | longitud(lista) == 0 = False
--                          | otherwise = (head lista) == elemento || pertenece (tail lista) elemento

hayQueCodificar :: Char -> [(Char, Char)] -> Bool
hayQueCodificar caracter mapeo | longitud(mapeo) == 0 = False
                               | otherwise = caracter == (fst(head mapeo)) || hayQueCodificar caracter (tail mapeo)

{-Ejercicio 2 (2 puntos)
problema cuantasVecesHayQueCodificar (c: Char, frase: seq⟨Char⟩, mapeo: seq⟨Char x Char⟩ ) : Z {
  requiere: {No hay elementos repetidos entre las primeras componentes de mapeo}
  requiere: {No hay elementos repetidos entre las segundas componentes de mapeo}
  requiere: {|frase| > 0}
  requiere: {c pertenece a frase}
  asegura: {(res = 0 y hayQueCodificar (c, mapeo) = false) o 
            (res = cantidad de veces que c aparece en frase y hayQueCodificar (c, mapeo) = true)}
}
-}
cantApariciones :: [Char] -> Char -> Int
cantApariciones lista elemento | longitud(lista) == 0 = 0
                               | elemento == (head lista) = 1 + cantApariciones (tail lista) elemento
                               | otherwise = cantApariciones (tail lista) elemento

cuantasVecesHayQueCodificar :: Char -> [Char] -> [(Char, Char)] -> Int
cuantasVecesHayQueCodificar c frase mapeo | not(hayQueCodificar c mapeo) = 0
                                          | otherwise = cantApariciones frase c

{-Ejercicio 3 (2 puntos)
problema laQueMasHayQueCodificar (frase: seq⟨Char⟩, mapeo: seq⟨Char x Char⟩ ) : Char {
  requiere: {No hay elementos repetidos entre las primeras componentes de mapeo}
  requiere: {No hay elementos repetidos entre las segundas componentes de mapeo}
  requiere: {|frase| > 0}
  requiere: {Existe al menos un c que pertenece a frase y hayQueCodificar(c, mapeo) = true}
  asegura: {res = c donde c es el caracter tal que cuantasVecesHayQueCodificar(c, frase, mapeo) es mayor a cualquier otro caracter
            perteneciente a frase}
  asegura: {Si existen más de un caracter c que cumple la condición anterior, devuelve el que aparece primero en frase}
}
-}
caracterQueMasHayQueCodificar :: [Char] -> [Char] -> [(Char, Char)] -> Char -> Char
caracterQueMasHayQueCodificar frase mismaFrase mapeo mayor | longitud(frase) == 0 = mayor
                                                           | cuantasVecesHayQueCodificar (head frase) mismaFrase mapeo > cuantasVecesHayQueCodificar mayor mismaFrase mapeo = caracterQueMasHayQueCodificar (tail frase) mismaFrase mapeo (head frase)
                                                           | otherwise = caracterQueMasHayQueCodificar (tail frase) mismaFrase mapeo mayor

laQueMasHayQueCodificar :: [Char] -> [(Char, Char)] -> Char
laQueMasHayQueCodificar frase mapeo = caracterQueMasHayQueCodificar frase frase mapeo (head frase)

{-Ejercicio 4 (3 puntos)
problema codificarFrase (frase: seq⟨Char⟩, mapeo: seq⟨Char x Char⟩ ) : seq ⟨Char⟩ {
  requiere: {No hay elementos repetidos entre las primeras componentes de mapeo}
  requiere: {No hay elementos repetidos entre las segundas componentes de mapeo}
  requiere: {|frase| > 0 }
  asegura: {|res| = |frase|}
  asegura: { Para todo 0 <= i < |frase| si hayQueCodificar(frase[i], mapeo) = true entonces
             res[i] = mapeo[j]_1, para un j tal que 0 <= j < |mapeo| y mapeo[j]_0 = frase[i]}
  asegura: { Para todo 0 <= i < |frase| si hayQueCodificar(frase[i], mapeo) = false entonces res[i] = frase[i]}
}
-}
buscoElMapeo :: Char -> [(Char, Char)] -> Char
buscoElMapeo caracter mapeo | fst(head mapeo) == caracter = snd(head mapeo)
                            | otherwise = buscoElMapeo caracter (tail mapeo)
                            
codificarFrase :: [Char] -> [(Char, Char)] -> [Char]
codificarFrase frase mapeo | longitud(frase) == 0 = []
                           | hayQueCodificar (head frase) mapeo = (buscoElMapeo (head frase) mapeo):codificarFrase (tail frase) mapeo
                           | otherwise = (head frase):codificarFrase (tail frase) mapeo

{-Ejercicio 5 (1 punto) 
Conteste marcando la opción correcta.
Si un usuario no cumple con la precondición de la especificación de un programa y el programa no termina (se cuelga):
1) El usuario tiene derecho a quejarse porque el programador debería haber contemplado ese caso.
2) El usuario no tiene derecho a quejarse, pero el programa es incorrecto porque no debería colgarse.
3) El usuario no tiene derecho a quejarse y no importa que el programa se cuelgue para este caso.
-}
--Ejercicio 5: El usuario no tiene derecho a quejarse y no importa que el programa se cuelgue para este caso.